src/AspNetCoreModuleV2/InProcessRequestHandler/inprocessapplication.cpp

```diff
@@ -35,161 +29,59 @@ IN_PROCESS_APPLICATION::IN_PROCESS_APPLICATION(
-            QueryConfigPath().c_str());
-    }
-
+{   
+    StopClr();
     InProcessApplicationBase::StopInternal(fServerInitiated);
 }
 
 VOID
-IN_PROCESS_APPLICATION::ShutDownInternal()
-{
-    DWORD    dwThreadStatus = 0;
-    DWORD    dwTimeout = m_pConfig->QueryShutdownTimeLimitInMS();
-
-    if (IsDebuggerPresent())
-    {
-        dwTimeout = INFINITE;
-    }
-
-    if (m_fShutdownCalledFromNative ||
-        m_status == MANAGED_APPLICATION_STATUS::STARTING ||
-        m_status == MANAGED_APPLICATION_STATUS::FAIL)
-    {
-        return;
-    }
-
+IN_PROCESS_APPLICATION::StopClr()
+{   
+    LOG_INFO(L"Stopping CLR");
+
+    // We need to keep track of when both managed and native initiate shutdown
+    // to avoid AVs. If shutdown has already been initiated in managed, we don't want to call into
+    // managed. We still need to wait on main exiting no matter what. m_fShutdownCalledFromNative
+    // is used for detecting redundant calls and blocking more requests to OnExecuteRequestHandler.
+    
+    if (!m_blockManagedCallbacks)
     {
-        if (m_fShutdownCalledFromNative ||
-            m_status == MANAGED_APPLICATION_STATUS::STARTING ||
-            m_status == MANAGED_APPLICATION_STATUS::FAIL)
-        {
-            return;
-        }
+        // We cannot call into managed if the dll is detaching from the process.
+        // Calling into managed code when the dll is detaching is strictly a bad idea,
+        // and usually results in an AV saying "The string binding is invalid"
+        const auto shutdownHandler = m_ShutdownHandler;
 
-        // We need to keep track of when both managed and native initiate shutdown
-        // to avoid AVs. If shutdown has already been initiated in managed, we don't want to call into
-        // managed. We still need to wait on main exiting no matter what. m_fShutdownCalledFromNative
-        // is used for detecting redundant calls and blocking more requests to OnExecuteRequestHandler.
-        m_fShutdownCalledFromNative = TRUE;
-        m_status = MANAGED_APPLICATION_STATUS::SHUTDOWN;
-
-        if (!m_fShutdownCalledFromManaged)
+        if (!g_fProcessDetach && shutdownHandler != nullptr)
         {
-            // We cannot call into managed if the dll is detaching from the process.
-            // Calling into managed code when the dll is detaching is strictly a bad idea,
-            // and usually results in an AV saying "The string binding is invalid"
-            if (!g_fProcessDetach)
-            {
-                m_ShutdownHandler(m_ShutdownHandlerContext);
-                m_ShutdownHandler = NULL;
-            }
+            shutdownHandler(m_ShutdownHandlerContext);
         }
+    }
 
-        // Release the lock before we wait on the thread to exit.
+    // Signal shutdown
+    if (m_pShutdownEvent != nullptr)
+    {
+        LOG_IF_FAILED(SetEvent(m_pShutdownEvent));
     }
 
-    if (!m_fShutdownCalledFromManaged)
+    if (m_workerThread.joinable())
```

Comment:
When would this ever be false?

Reply:
If the thread wasn't initialized.

Reply:
Check what for null?

Reply:
if ` m_workerThread != nullptr`

Reply:
m_workerThread is not a reference. It can't be null.